url53f943c5cb_Praznat_AGP_tgz-pJ8-AMath_TestingJ8/src/Game/Contract.java

```diff
@@ -0,0 +1,257 @@
+		final double offerValue = getOfferValue();
+		accepted = demandValue + offerValue > 0;
+		if (!AGPmain.AUTOPILOT && AGPmain.mainGUI != null && evaluator == AGPmain.mainGUI.AC.getAvatar()) {avatarChooseAcceptable(accepted);}
+		evaluator.addReport(GobLog.contractOutcome(proposer, evaluator, accepted, demandValue, offerValue));
+		proposer.addReport(GobLog.contractOutcome(proposer, evaluator, accepted, demandValue, offerValue));
+		return accepted;
+	}
+
+	public double getPreCalcedDemandValue() {return demandValue;}
+	public double getPreCalcedOfferValue() {return offerValue;}
+	public double getDemandValue() {
+		demandValue = 0;
+		for (DealTerm dt : demands) {demandValue += dt.getEvaluation();}
+		return demandValue;
+	}
+	public double getOfferValue() {
+		offerValue = 0;
+		for (DealTerm dt : offers) {offerValue += dt.getEvaluation();}
+		return offerValue;
+	}
+	
+	private void avatarChooseAcceptable(boolean aiWouldAccept) {
+		String prompt = "";
+		for (DealTerm dt : demands) {prompt += dt.loggy() + ";";}
+		for (DealTerm dt : offers) {prompt += dt.loggy() + ";";}
+		final Boolean[] choices = aiWouldAccept ? new Boolean[] {Boolean.TRUE, Boolean.FALSE} : new Boolean[] {Boolean.FALSE, Boolean.TRUE};
+		AGPmain.mainGUI.AC.showChoices(prompt, evaluator, choices, SubjectiveType.NO_ORDER, new Calc.Listener() {
+			@Override
+			public void call(Object arg) {
+				accepted = ((Boolean)arg).booleanValue();
+			}
+		}, new Calc.Transformer<Boolean, String>() {
+			@Override
+			public String transform(Boolean b) {
+				return b.booleanValue() ? "Accept terms" : "Reject terms";
+			}
+		});
+	}
+	public void enact() {
+		for (DealTerm offer : offers) {if (!offer.doit()) {evaluator.AB.add(new Stressor(Stressor.PROMISE_BROKEN, proposer));}}
+		for (DealTerm demand : demands) {if (!demand.doit()) {proposer.AB.add(new Stressor(Stressor.PROMISE_BROKEN, evaluator));}}
+	}
+
+	private double milletEval(int millet) {
+		return ((Assessable)Values.WEALTH).evaluate(evaluator, null, millet);
+	}
+	
+	protected abstract class DealTerm {
+		public double getEvaluation() {
+			final Reportable loggy = loggy();
+			evaluator.addReport(loggy); proposer.addReport(loggy); return evaluate();}
+		protected abstract double evaluate();
+		protected abstract boolean doit();
+		protected abstract Reportable loggy();
+	}
+	public void demandTribute(int millet) {demands.add(new DemandTributeTerm(millet));}
+	protected class DemandTributeTerm extends DealTerm {
+		private final int millet;
+		public DemandTributeTerm(int millet) {this.millet = millet;}
+		@Override
+		protected double evaluate() {
+			return milletEval(-Math.min(millet, evaluator.getMillet())); // evaluator never assumes debt
+		}
+		@Override
+		protected boolean doit() {
+			boolean enough = evaluator.getMillet() >= millet;
+			int amt = (enough ? millet : evaluator.getMillet());
+			evaluator.alterMillet(-amt);
+			proposer.alterMillet(amt);
+			return enough;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 108:112, 176:180.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.